homeassistant/components/onvif/camera.py

```diff
@@ -16,142 +15,95 @@
-    _LOGGER.debug("Setting up the ONVIF camera platform")
-
-    async def async_handle_ptz(service):
-        """Handle PTZ service call."""
-        pan = service.data.get(ATTR_PAN)
-        tilt = service.data.get(ATTR_TILT)
-        zoom = service.data.get(ATTR_ZOOM)
-        distance = service.data[ATTR_DISTANCE]
-        speed = service.data[ATTR_SPEED]
-        move_mode = service.data.get(ATTR_MOVE_MODE)
-        continuous_duration = service.data[ATTR_CONTINUOUS_DURATION]
-        preset = service.data[ATTR_PRESET]
-        all_cameras = hass.data[ONVIF_DATA][ENTITIES]
-        entity_ids = await async_extract_entity_ids(hass, service)
-        target_cameras = []
-        if not entity_ids:
-            target_cameras = all_cameras
-        else:
-            target_cameras = [
-                camera for camera in all_cameras if camera.entity_id in entity_ids
-            ]
-        for camera in target_cameras:
-            await camera.async_perform_ptz(
-                pan, tilt, zoom, distance, speed, move_mode, continuous_duration, preset
-            )
-
-    hass.services.async_register(
-        DOMAIN, SERVICE_PTZ, async_handle_ptz, schema=SERVICE_PTZ_SCHEMA
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the ONVIF camera video stream."""
+    platform = entity_platform.current_platform.get()
+
+    # Create PTZ service
+    platform.async_register_entity_service(
+        SERVICE_PTZ,
+        {
+            vol.Optional(ATTR_PAN): vol.In([DIR_LEFT, DIR_RIGHT]),
+            vol.Optional(ATTR_TILT): vol.In([DIR_UP, DIR_DOWN]),
+            vol.Optional(ATTR_ZOOM): vol.In([ZOOM_OUT, ZOOM_IN]),
+            vol.Optional(ATTR_DISTANCE, default=0.1): cv.small_float,
+            vol.Optional(ATTR_SPEED, default=0.5): cv.small_float,
+            vol.Optional(ATTR_MOVE_MODE, default=RELATIVE_MOVE): vol.In(
+                [CONTINUOUS_MOVE, RELATIVE_MOVE, ABSOLUTE_MOVE, GOTOPRESET_MOVE]
+            ),
+            vol.Optional(ATTR_CONTINUOUS_DURATION, default=0.5): cv.small_float,
+            vol.Optional(ATTR_PRESET, default="0"): cv.string,
+        },
+        "async_perform_ptz",
     )
 
-    _LOGGER.debug("Constructing the ONVIFHassCamera")
-
-    hass_camera = ONVIFHassCamera(hass, config)
+    base_config = {
+        CONF_NAME: config_entry.data[CONF_NAME],
+        CONF_HOST: config_entry.data[CONF_HOST],
+        CONF_PORT: config_entry.data[CONF_PORT],
+        CONF_USERNAME: config_entry.data[CONF_USERNAME],
+        CONF_PASSWORD: config_entry.data[CONF_PASSWORD],
+        CONF_EXTRA_ARGUMENTS: config_entry.options[CONF_EXTRA_ARGUMENTS],
+        CONF_RTSP_TRANSPORT: config_entry.options[CONF_RTSP_TRANSPORT],
+    }
 
-    await hass_camera.async_initialize()
+    entities = []
+    for profile in config_entry.data[CONF_PROFILE]:
+        config = {**base_config, CONF_PROFILE: profile}
+        camera = ONVIFHassCamera(hass, config)
+        await camera.async_initialize()
```

Comment:
Does the camera initialization need to happen sequentially? Could we use `asyncio.gather` to make it concurrent?